File: 2023/d04/day.rb
cards = File.readlines('input.txt').map {|l| l.strip.split(":")}
acc = Hash.new{|h,k| h[k]=0}

p1 = cards.map.with_index do |(card_number, numbers), i|
  acc[i] += 1
  winning_numbers, my_numbers = numbers.split("|").map {|l| l.strip.split(" ").map(&:to_i)}
  matches = (winning_numbers & my_numbers).count

  matches.times do |j|
    acc[i + 1 + j] += acc[i]
  end
  matches > 0 ? 2 ** (matches - 1) : 0
end.sum

puts p1
puts acc.values.sum

File: 2023/d04/day.exs
day =
  File.read!("input.txt")
  |> String.split("\n", trim: true)
  |> Enum.with_index()
  |> Enum.reduce(%{part1: %{}}, fn {line, i}, acc ->
    acc =
      acc
      |> Map.put(i, Map.get(acc, i, 0) + 1)

    [_, numbers] = String.split(line, ":", trim: true)

    [winning_numbers, my_numbers] =
      numbers
      |> String.split("|")
      |> Enum.map(fn list ->
        String.split(list, " ", trim: true)
        |> Enum.map(&String.to_integer(&1))
        |> MapSet.new()
      end)

    matches =
      MapSet.intersection(winning_numbers, my_numbers)
      |> Enum.count()

    if matches > 0 do
      acc =
        for j <- 1..matches, reduce: acc do
          acc ->
            current_val = Map.get(acc, i)
            Map.put(acc, i + j, current_val + Map.get(acc, i + j, 0))
        end

      put_in(acc, [:part1, i], 2 ** (matches - 1))
    else
      put_in(acc, [:part1, i], 0)
    end
  end)

Map.get(day, :part1) |> Map.values() |> Enum.sum() |> IO.puts()
Map.delete(day, :part1) |> Map.values() |> Enum.sum() |> IO.puts()
